### PR TITLE
batch get

### PR DESCRIPTION
**Pull request by morefreeze**
PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @disksing @qiuyesuifeng

File: snapshot.go [hunk 1]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+func (s *tikvSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
+	m := make(map[string][]byte, len(keys))
+
+	multiBatchGet, err := s.makeBatchGetReqs(keys)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, singleBatchGet := range multiBatchGet {
+		keys := singleBatchGet.GetKeys()
+		for startIdx := 0; startIdx < len(keys); startIdx += int(batchGetSize) {
```
Comment (disksing) [inline snapshot.go]: You may define const value `batchGetSize` without specific type. Then you don't have to do the casting here.

**Comment by disksing** [comment]
The rest LGTM.

File: snapshot.go [hunk 2]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+}
+
+// doBatchGet sends BatchGet RPC. If some key is locked calling Get directly.
+func (s *tikvSnapshot) doBatchGet(singleBatchGet *batchGetRegion) (map[string][]byte, error) {
+	cmdBatchGetReq := singleBatchGet.CmdBatchGetRequest
+	keys := cmdBatchGetReq.GetKeys()
+	if len(keys) == 0 {
+		return nil, nil
+	}
+	// go func()
```
Comment (qiuyesuifeng) [inline snapshot.go]: What is the meaning of this comment?

File: region_cache_test.go [hunk 3]
```diff
@@ -72,7 +73,7 @@ func (s *testRegionCacheSuite) TestDropRegion(c *C) {
 
 func (s *testRegionCacheSuite) TestDropRegion2(c *C) {
 	// remove store 1 from PD, should cause an error in `loadRegion()`
-	s.pd.removeStore(1)
+	s.pd.removeStore(2)
```
Comment (qiuyesuifeng) [inline region_cache_test.go]: Above comment should be updated.

File: snapshot.go [hunk 4]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+			fallthrough
+		default:
+			return nil, errors.Errorf("unknown error: %s", resType)
 		}
 	}
 	return m, nil
 }
 
+// BatchGet return map[key] => value, if keyNotExist map won't contain this key.
+func (s *tikvSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
```
Comment (qiuyesuifeng) [inline snapshot.go]: Seems we can send batch requests by region id parallelly later?

File: snapshot.go [hunk 5]
```diff
@@ -149,7 +247,26 @@ func (s *tikvSnapshot) handleGetError(resType *pb.ResultType, key []byte) ([]byt
 	return nil, errors.Errorf("unknown error: %s", resType)
 }
 
+// addDict add d2 to d1, if d2 has any key of d1 return error.
+func addDict(d1, d2 map[string][]byte) (map[string][]byte, error) {
```
Comment (qiuyesuifeng) [inline snapshot.go]: Can u show an example for key conflict?

File: snapshot.go [hunk 6]
```diff
@@ -27,8 +27,10 @@ var (
 )
 
 const (
-	tikvBatchSize = 100
-	maxGetCount   = 3
+	tikvBatchSize    = 100
```
Comment (shenli) [inline snapshot.go]: What's the difference between batchGetSize and tikvBatchSize?

File: snapshot_test.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	// Clean all data.
+	txn := s.beginTxn(c)
+	scanner, err := txn.Seek(encodeKey(""))
+	c.Assert(err, IsNil)
+	c.Assert(scanner, NotNil)
+	for scanner.Valid() {
+		k := scanner.Key()
+		err := txn.Delete(k)
+		c.Assert(err, IsNil)
+		scanner.Next()
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: We should check all return error in tests.

File: snapshot.go [hunk 8]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
 	}
 }
 
-// BatchGet implements kv.Snapshot.BatchGet interface.
-func (s *tikvSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
-	// TODO: send batch get rpc
-	m := make(map[string][]byte, len(keys))
+// makeBatchGetReqs split each key into corresponding region.
```
Comment (shenli) [inline snapshot.go]: split->splits

File: snapshot.go (hunk 2 above)
Comment (morefreeze) [inline snapshot.go]: Check results can use go func() concurrently. We need confirm correctness.

File: region_cache_test.go (hunk 3 above)
Comment (morefreeze) [inline region_cache_test.go]: Got it.

File: snapshot.go (hunk 4 above)
Comment (morefreeze) [inline snapshot.go]: I will add comment about it.

File: snapshot_test.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	c.Assert(err, IsNil)
+	cnt := 0
+	for {
+		if !scan.Valid() {
+			break
+		}
+		cnt++
+		k := scan.Key()
+		v := scan.Value()
+		c.Assert(v, BytesEquals, m[string(k)])
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: Seems `m[string(k)]` has two return values, can we directly using `BytesEquals`?

File: snapshot.go (hunk 5 above)
Comment (morefreeze) [inline snapshot.go]: Sure.

File: snapshot_test.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	scan.Next()
+	c.Assert(scan.Valid(), IsFalse)
+	err = txn.Commit()
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, cnt)
+}
+
+func (s *testSnapshotSuite) TestBatchGet(c *C) {
+	rowNum := 200
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: It is better to add corner case test, like `rowNum  = batchSize * 2 + 1` or `rowNum  = batchSize * 2 - 1`?

File: snapshot.go (hunk 5 above)
Comment (shenli) [inline snapshot.go]: mergeResult is a better name?

File: snapshot_test.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	c.Assert(err, IsNil)
+
+	keys := makeKeys(rowNum)
+	txn3 := s.beginTxn(c)
+	snapshot := newTiKVSnapshot(s.store, kv.Version{Ver: txn3.tid})
+	_, err = snapshot.BatchGet(keys)
+	// rowNum/2 key is locked, failed
+	c.Assert(err, NotNil)
+
+	time.Sleep(3e9)
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: Can we get stable result if u sleep such a short time?

File: snapshot.go (hunk 2 above)
Comment (morefreeze) [inline snapshot.go]: My mistake, this comment is useless.

File: snapshot.go [hunk 12]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+		return nil, errors.Trace(errBodyMissing)
+	}
+	if resType := cmdBatchGetResp.GetResType(); resType == nil || resType.GetType() != pb.ResultType_Ok {
+		log.Errorf("Unexpected result type [%s]", resType)
+		return nil, errors.Trace(errRetryable)
+	}
+	results := cmdBatchGetResp.GetResults()
+	log.Debugf("Receive batchGet results length[%d]", len(results))
+	m := make(map[string][]byte, len(results))
+	// go func()
```
Comment (shenli) [inline snapshot.go]: Useless comment?

File: snapshot.go [hunk 13]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+			multiBatchGet[regionID] = singleBatchGet
+		}
+		cmdBatchGetReq := singleBatchGet.CmdBatchGetRequest
+		cmdBatchGetReq.Keys = append(cmdBatchGetReq.Keys, k)
+		cmdBatchGetReq.Version = proto.Uint64(startTS)
+	}
+	return multiBatchGet, nil
+}
+
+// doBatchGet sends BatchGet RPC. If some key is locked calling Get directly.
```
Comment (shenli) [inline snapshot.go]: doBatchGet sends BatchGet RPC request. If any key is locked, use tikvSnapshot.Get() to retry.

File: snapshot.go [hunk 14]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+			log.Errorf("Be serious, it can't happen here. Server may make a joke.")
+			fallthrough
+		default:
+			return nil, errors.Errorf("unknown error: %s", resType)
 		}
 	}
 	return m, nil
 }
 
+// BatchGet return map[key] => value, if keyNotExist map won't contain this key.
```
Comment (shenli) [inline snapshot.go]: BatchGet gets all the keys's value from kv-server and returns a map contains key/value pairs. The map will not contain nonexistent keys.

File: snapshot.go (hunk 5 above)
Comment (shenli) [inline snapshot.go]: Merge d2 into d1. If d1 and d2 are overlap, it returns error.

Comment (morefreeze) [inline snapshot.go]: Better.

File: snapshot.go (hunk 6 above)
Comment (morefreeze) [inline snapshot.go]: tikvBatchSize => scanBatchSize

File: snapshot_test.go (hunk 7 above)
Comment (morefreeze) [inline snapshot_test.go]: I don't know how many key in table, so I don't know when Next return error.

File: snapshot_test.go (hunk 10 above)
Comment (morefreeze) [inline snapshot_test.go]: How about add random rowNum?

File: snapshot_test.go (hunk 11 above)
Comment (morefreeze) [inline snapshot_test.go]: I try to fix this.

File: snapshot.go (hunk 12 above)
Comment (morefreeze) [inline snapshot.go]: change concurrent later.

File: snapshot.go (hunk 14 above)
Comment (morefreeze) [inline snapshot.go]: all the keys' value?

File: snapshot_test.go (hunk 9 above)
Comment (disksing) [inline snapshot_test.go]: @qiuyesuifeng It could has either 1 or 2 return values, compiler will infer it automatically.

File: snapshot.go [hunk 15]
```diff
@@ -45,22 +47,117 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+		if !ok {
+			singleBatchGet = &batchGetRegion{
+				CmdBatchGetRequest: &pb.CmdBatchGetRequest{},
+				region:             region,
+			}
+			multiBatchGet[regionID] = singleBatchGet
+		}
+		cmdBatchGetReq := singleBatchGet.CmdBatchGetRequest
+		cmdBatchGetReq.Keys = append(cmdBatchGetReq.Keys, k)
+		cmdBatchGetReq.Version = proto.Uint64(startTS)
```
Comment (coocood) [inline snapshot.go]: Set `Version` when creating CmdBatchGetRequest.

Comment (morefreeze) [inline snapshot.go]: Got it.

File: snapshot_test.go [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+		for i := 0; i < rowNum; i++ {
+			k := encodeKey(fmt.Sprintf("key%08d", i))
+			err := txn.Set(k, []byte(fmt.Sprintf("value%08d", i)))
+			c.Assert(err, IsNil)
+		}
+		err := txn.Commit()
+		c.Assert(err, IsNil)
+
+		txn2 := s.beginTxn(c)
+		txn2.DONOTCOMMIT = true
```
Comment (siddontang) [inline snapshot_test.go]: DONOTCOMMIT is not go style.

File: snapshot_test.go [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+func makeDict(keys []string) map[string][]byte {
+	d := make(map[string][]byte)
+	for _, k := range keys {
+		d[k] = []byte(k)
+	}
+	return d
+}
+
+func equalByteDict(lhs, rhs map[string][]byte, c *C) {
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: Please using c as first variable.

File: snapshot_test.go [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+		}
+		cnt++
+		k := scan.Key()
+		v := scan.Value()
+		v2, ok := m[string(k)]
+		c.Assert(ok, IsTrue, Commentf("key: %q", k))
+		c.Assert(v, BytesEquals, v2)
+		scan.Next()
+	}
+	scan.Next()
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: Why here call Next again?

File: snapshot_test.go [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+
+		keys := makeKeys(rowNum)
+		s.checkAll(keys, c)
+		s.deleteKeys(keys, c)
+	}
+}
+
+func (s *testSnapshotSuite) TestBatchGetLock(c *C) {
+	for i := 0; i < maxBatchGet; i++ {
+		rowNum := rand.Intn(1000)
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: TestBatchGetLock

File: snapshot_test.go [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+		time.Sleep(time.Duration(cleanupTimeout) * time.Millisecond)
+		s.checkAll(keys, c)
+		s.deleteKeys(keys, c)
+	}
+}
+
+func (s *testSnapshotSuite) TestBatchGetNotExist(c *C) {
+	for i := 0; i < maxBatchGet; i++ {
+		rowNum := rand.Intn(1000)
+		log.Debugf("Test BatchGet with length[%d]", rowNum)
```
Comment (qiuyesuifeng) [inline snapshot_test.go]: TestBatchGetNotExist

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
Please squash your commit log.
